File: tpcp-rs/tpcp-std/src/node.rs
use std::collections::HashMap;
use std::sync::Arc;
use std::sync::atomic::{AtomicUsize, Ordering};
use tokio::sync::RwLock;
use futures_util::{SinkExt, StreamExt};
use tokio_tungstenite::{connect_async, accept_async, tungstenite::Message};
use serde_json;
use tpcp_core::{AgentIdentity, Intent, TPCPEnvelope, MessageHeader, PROTOCOL_VERSION};
use tpcp_core::identity::{canonical_json, sign, verify};
use ed25519_dalek::SigningKey;
use uuid::Uuid;
use crate::DLQ;

type HandlerFn = Arc<dyn Fn(TPCPEnvelope) + Send + Sync + 'static>;

/// Async TPCP node backed by tokio and tokio-tungstenite.
pub struct TPCPNode {
    pub identity: AgentIdentity,
    pub memory: Arc<RwLock<tpcp_core::LWWMap>>,
    pub dlq: Arc<DLQ>,
    handlers: Arc<RwLock<HashMap<Intent, HandlerFn>>>,
    /// Optional Ed25519 signing key. When present, `send_message` attaches a signature.
    signing_key: Option<Arc<SigningKey>>,
    /// Known peer public keys: agent_id → base64-encoded public key.
    /// Inbound messages from registered peers are signature-verified before dispatch.
    peer_keys: Arc<RwLock<HashMap<String, String>>>,
    /// Replay-protection window: message_id → unix timestamp (seconds) of first receipt.
    seen_messages: Arc<RwLock<HashMap<String, i64>>>,
    /// Maximum number of concurrent inbound WebSocket connections.
    max_connections: usize,
    /// Current count of active inbound connections.
    active_connections: Arc<AtomicUsize>,
}

impl TPCPNode {
    /// Creates a new TPCPNode with the given identity and no signing key.
    pub fn new(identity: AgentIdentity) -> Self {
        Self {
            identity,
            memory: Arc::new(RwLock::new(tpcp_core::LWWMap::new())),
            dlq: Arc::new(DLQ::new()),
            handlers: Arc::new(RwLock::new(HashMap::new())),
            signing_key: None,
            peer_keys: Arc::new(RwLock::new(HashMap::new())),
            seen_messages: Arc::new(RwLock::new(HashMap::new())),
            max_connections: 100,
            active_connections: Arc::new(AtomicUsize::new(0)),
        }
    }

    /// Creates a TPCPNode that signs all outbound messages.
    pub fn with_signing_key(identity: AgentIdentity, key: SigningKey) -> Self {
        Self {
            identity,
            memory: Arc::new(RwLock::new(tpcp_core::LWWMap::new())),
            dlq: Arc::new(DLQ::new()),
            handlers: Arc::new(RwLock::new(HashMap::new())),
            signing_key: Some(Arc::new(key)),
            peer_keys: Arc::new(RwLock::new(HashMap::new())),
            seen_messages: Arc::new(RwLock::new(HashMap::new())),
            max_connections: 100,
            active_connections: Arc::new(AtomicUsize::new(0)),
        }
    }

    /// Registers a peer's public key for inbound signature verification.
    pub async fn register_peer_key(&self, agent_id: &str, public_key_b64: &str) {
        self.peer_keys.write().await.insert(agent_id.to_string(), public_key_b64.to_string());
    }

    /// Registers a handler for a specific intent.
    pub async fn register_handler<F>(&self, intent: Intent, handler: F)
    where
        F: Fn(TPCPEnvelope) + Send + Sync + 'static,
    {
        self.handlers.write().await.insert(intent, Arc::new(handler));
    }

    /// Connects as a WebSocket client to the given URL.
    pub async fn connect(&self, url: &str) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let (ws_stream, _) = connect_async(url).await?;
        let (_, mut read) = ws_stream.split();
        let handlers = Arc::clone(&self.handlers);
        let dlq = Arc::clone(&self.dlq);
        let peer_keys = Arc::clone(&self.peer_keys);
        let seen_messages = Arc::clone(&self.seen_messages);
        tokio::spawn(async move {
            while let Some(Ok(msg)) = read.next().await {
                if let Message::Text(text) = msg {
                    if let Ok(env) = serde_json::from_str::<TPCPEnvelope>(&text) {
                        Self::dispatch_env(env, &handlers, &dlq, &peer_keys, &seen_messages).await;
                    }
                }
            }
        });
        Ok(())
    }

    /// Starts listening on the given address (e.g. "127.0.0.1:9001").
    pub async fn listen(&self, addr: &str) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let listener = tokio::net::TcpListener::bind(addr).await?;
        let handlers = Arc::clone(&self.handlers);
        let dlq = Arc::clone(&self.dlq);
        let peer_keys = Arc::clone(&self.peer_keys);
        let seen_messages = Arc::clone(&self.seen_messages);
        let max_connections = self.max_connections;
        let active_connections = Arc::clone(&self.active_connections);
        tokio::spawn(async move {
            while let Ok((stream, _)) = listener.accept().await {
                // Enforce connection limit before upgrading to WebSocket.
                if active_connections.load(Ordering::Acquire) >= max_connections {
                    eprintln!("[TPCPNode] max_connections ({}) reached — rejecting new connection", max_connections);
                    // Drop the raw TcpStream; this closes the TCP connection.
                    drop(stream);
                    continue;
                }

                let ws = match accept_async(stream).await {
                    Ok(ws) => ws,
                    Err(_) => continue,
                };
                let (_, mut read) = ws.split();
                let handlers = Arc::clone(&handlers);
                let dlq = Arc::clone(&dlq);
                let peer_keys = Arc::clone(&peer_keys);
                let seen_messages = Arc::clone(&seen_messages);
                let active_connections_inner = Arc::clone(&active_connections);
                active_connections_inner.fetch_add(1, Ordering::AcqRel);
                tokio::spawn(async move {
                    while let Some(Ok(msg)) = read.next().await {
                        if let Message::Text(text) = msg {
                            if let Ok(env) = serde_json::from_str::<TPCPEnvelope>(&text) {
                                Self::dispatch_env(env, &handlers, &dlq, &peer_keys, &seen_messages).await;
                            }
                        }
                    }
                    active_connections_inner.fetch_sub(1, Ordering::AcqRel);
                });
            }
        });
        Ok(())
    }

    /// Sends a message to the given WebSocket URL.
    /// `receiver_id` is the target agent's agent_id (not the URL).
    pub async fn send_message(
        &self,
        url: &str,
        receiver_id: &str,
        intent: Intent,
        payload: serde_json::Value,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let (mut ws_stream, _) = connect_async(url).await?;
        let signature = self.signing_key.as_ref().map(|key| {
            let canonical = canonical_json(&payload);
            sign(key, &canonical)
        });
        let envelope = TPCPEnvelope {
            header: MessageHeader {
                message_id: uuid_v4(),
                sender_id: self.identity.agent_id.clone(),
                receiver_id: receiver_id.to_string(),
                intent,
                timestamp: chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
                ttl: 30,
                protocol_version: PROTOCOL_VERSION.to_string(),
            },
            payload,
            signature,
            ack_info: None,
            chunk_info: None,
        };
        let text = serde_json::to_string(&envelope)?;
        ws_stream.send(Message::Text(text)).await?;
        ws_stream.close(None).await?;
        Ok(())
    }

    async fn dispatch_env(
        env: TPCPEnvelope,
        handlers: &RwLock<HashMap<Intent, HandlerFn>>,
        dlq: &DLQ,
        peer_keys: &RwLock<HashMap<String, String>>,
        seen_messages: &RwLock<HashMap<String, i64>>,
    ) {
        // --- Signature enforcement ---
        // If the envelope carries a non-empty signature, the sender's public key
        // MUST be registered. Fail closed: unknown-peer signed messages go to DLQ.
        let has_signature = env.signature.as_deref().map(|s| !s.is_empty()).unwrap_or(false);
        let sender_key = peer_keys.read().await.get(&env.header.sender_id).cloned();

        if has_signature && sender_key.is_none() {
            eprintln!(
                "[TPCPNode] signed envelope from unknown peer '{}' (message_id: {}) — routing to DLQ",
                env.header.sender_id, env.header.message_id
            );
            if !dlq.enqueue(env) {
                eprintln!("[TPCPNode] DLQ full — dropping signed-unknown-peer envelope");
            }
            return;
        }

        if let Some(pub_key_b64) = sender_key {
            let canonical = canonical_json(&env.payload);
            let sig_ok = env.signature.as_deref()
                .map(|sig| verify(&pub_key_b64, &canonical, sig))
                .unwrap_or(false);
            if !sig_ok {
                // Invalid or missing signature from a known peer — route to DLQ.
                if !dlq.enqueue(env) {
                    eprintln!("[TPCPNode] DLQ full — dropping invalid-signature envelope");
                }
                return;
            }
        }

        // --- Replay protection ---
        let now = chrono::Utc::now().timestamp();
        let message_id = env.header.message_id.clone();
        {
            let mut seen = seen_messages.write().await;
            // Evict entries older than 300 seconds.
            seen.retain(|_, &mut ts| now - ts < 300);
            // Check for duplicate.
            if seen.contains_key(&message_id) {
                // Silently drop replayed messages.
                return;
            }
            // Record first-seen timestamp.
            seen.insert(message_id, now);
        }

        let handler = handlers.read().await.get(&env.header.intent).cloned();
        match handler {
            Some(h) => h(env),
            None => {
                if !dlq.enqueue(env) {
                    eprintln!("[TPCPNode] DLQ full — dropping unhandled envelope");
                }
            }
        }
    }
}

fn uuid_v4() -> String {
    Uuid::new_v4().to_string()
}

File: tpcp-go/tpcp/node.go
package tpcp

import (
	"crypto/ed25519"
	"encoding/json"
	"fmt"
	"log"
	"net"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// defaultUpgrader is kept for non-node uses; TPCPNode uses a per-node upgrader
// built in handleUpgrade to enforce AllowedOrigins.
var defaultUpgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool { return false },
}

// Validatable is implemented by payload types that can self-validate.
type Validatable interface {
	Validate() error
}

// TPCPNode is a TPCP agent node that can send and receive messages over WebSocket.
type TPCPNode struct {
	Identity *AgentIdentity
	PrivKey  ed25519.PrivateKey
	Memory   *LWWMap
	DLQ      *DLQ

	// AllowedOrigins is the list of permitted WebSocket origins for browser clients.
	// If empty, all origins are rejected when an Origin header is present.
	// Non-browser clients (no Origin header) are always allowed.
	AllowedOrigins []string

	// Ready is closed by Listen() once the server is bound and ready to accept connections.
	Ready chan struct{}

	handlers   map[Intent]func(*TPCPEnvelope)
	handlersMu sync.RWMutex

	// peerKeys maps agent_id → base64 public key for inbound signature verification.
	peerKeys   map[string]string
	peerKeysMu sync.RWMutex

	peers   map[string]*websocket.Conn
	peersMu sync.RWMutex

	seenMessages   map[string]int64
	seenMessagesMu sync.Mutex

	server *http.Server
	done   chan struct{}
	wg     sync.WaitGroup
}

// NewTPCPNode creates a TPCPNode with the given identity and optional private key.
func NewTPCPNode(identity *AgentIdentity, privKey ed25519.PrivateKey) *TPCPNode {
	return &TPCPNode{
		Identity: identity,
		PrivKey:  privKey,
		Memory:   NewLWWMap(),
		DLQ:      NewDLQ(),
		Ready:    make(chan struct{}),
		handlers: make(map[Intent]func(*TPCPEnvelope)),
		peerKeys:     make(map[string]string),
		peers:        make(map[string]*websocket.Conn),
		seenMessages: make(map[string]int64),
		done:         make(chan struct{}),
	}
}

// RegisterHandler registers a handler function for a specific intent.
func (n *TPCPNode) RegisterHandler(intent Intent, handler func(*TPCPEnvelope)) {
	n.handlersMu.Lock()
	defer n.handlersMu.Unlock()
	n.handlers[intent] = handler
}

// RegisterPeerKey registers the public key for a known peer so inbound messages can be verified.
func (n *TPCPNode) RegisterPeerKey(agentID, pubKeyB64 string) {
	n.peerKeysMu.Lock()
	defer n.peerKeysMu.Unlock()
	n.peerKeys[agentID] = pubKeyB64
}

// Listen starts a WebSocket server on the given address (e.g. ":8765").
// It binds the listener synchronously so that when Listen returns the port is
// already accepting connections. The Ready channel is closed immediately after
// the listener is bound.
func (n *TPCPNode) Listen(addr string) error {
	ln, err := net.Listen("tcp", addr)
	if err != nil {
		return fmt.Errorf("listen %s: %w", addr, err)
	}
	mux := http.NewServeMux()
	mux.HandleFunc("/", n.handleUpgrade)
	n.server = &http.Server{Addr: addr, Handler: mux}
	// Signal that the server is bound and ready to accept connections.
	close(n.Ready)
	n.wg.Add(1)
	go func() {
		defer n.wg.Done()
		if err := n.server.Serve(ln); err != nil && err != http.ErrServerClosed {
			log.Printf("[TPCPNode] server error: %v", err)
		}
	}()
	return nil
}

func (n *TPCPNode) handleUpgrade(w http.ResponseWriter, r *http.Request) {
	upgrader := websocket.Upgrader{
		CheckOrigin: func(req *http.Request) bool {
			origin := req.Header.Get("Origin")
			// Non-browser clients (no Origin header) are always allowed.
			if origin == "" {
				return true
			}
			// If no allowed origins are configured, deny all browser origins.
			if len(n.AllowedOrigins) == 0 {
				return false
			}
			for _, allowed := range n.AllowedOrigins {
				if origin == allowed {
					return true
				}
			}
			return false
		},
	}
	conn, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Printf("[TPCPNode] upgrade error: %v", err)
		return
	}
	peerID := r.RemoteAddr
	n.peersMu.Lock()
	n.peers[peerID] = conn
	n.peersMu.Unlock()
	n.wg.Add(1)
	go n.readLoop(peerID, conn)
}

// Connect establishes a WebSocket client connection to a peer URL.
func (n *TPCPNode) Connect(peerURL string) error {
	conn, _, err := websocket.DefaultDialer.Dial(peerURL, nil)
	if err != nil {
		return fmt.Errorf("connect to %s: %w", peerURL, err)
	}
	peerID := peerURL
	n.peersMu.Lock()
	n.peers[peerID] = conn
	n.peersMu.Unlock()
	n.wg.Add(1)
	go n.readLoop(peerID, conn)
	return nil
}

// SendMessage sends a message with the given intent and payload to a peer.
// peerID is the connection key (URL or remote address).
// receiverID is the agent_id of the target agent (used in the envelope header).
// payload must be JSON-serializable.
func (n *TPCPNode) SendMessage(peerID string, receiverID string, intent Intent, payload interface{}) error {
	if v, ok := payload.(Validatable); ok {
		if err := v.Validate(); err != nil {
			return fmt.Errorf("payload validation: %w", err)
		}
	}

	n.peersMu.RLock()
	conn, ok := n.peers[peerID]
	n.peersMu.RUnlock()
	if !ok {
		return fmt.Errorf("unknown peer: %s", peerID)
	}

	rawPayload, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("marshal payload: %w", err)
	}

	envelope := &TPCPEnvelope{
		Header: MessageHeader{
			MessageID:       randomUUID(),
			Timestamp:       time.Now().UTC().Format("2006-01-02T15:04:05.000Z"),
			SenderID:        n.Identity.AgentID,
			ReceiverID:      receiverID,
			Intent:          intent,
			TTL:             30,
			ProtocolVersion: PROTOCOL_VERSION,
		},
		Payload: rawPayload,
	}

	// Sign the payload with canonical JSON if we have a private key.
	if n.PrivKey != nil {
		canonical, err := CanonicalJSON(payload)
		if err != nil {
			return fmt.Errorf("canonical JSON for signing: %w", err)
		}
		envelope.Signature = Sign(n.PrivKey, canonical)
	}

	data, err := json.Marshal(envelope)
	if err != nil {
		return fmt.Errorf("marshal envelope: %w", err)
	}
	return conn.WriteMessage(websocket.TextMessage, data)
}

// Stop shuts down the WebSocket server and closes all peer connections.
// Connections are closed here and the peers map is replaced so that readLoop
// defers see an empty map and skip the redundant Close().
func (n *TPCPNode) Stop() error {
	close(n.done)
	// Close server FIRST so Serve() returns and the goroutine can exit.
	var serverErr error
	if n.server != nil {
		serverErr = n.server.Close()
	}
	n.peersMu.Lock()
	for _, conn := range n.peers {
		conn.Close()
	}
	n.peers = make(map[string]*websocket.Conn)
	n.peersMu.Unlock()
	n.wg.Wait()
	return serverErr
}

func (n *TPCPNode) readLoop(peerID string, conn *websocket.Conn) {
	defer n.wg.Done()
	defer func() {
		n.peersMu.Lock()
		if _, exists := n.peers[peerID]; exists {
			conn.Close()
			delete(n.peers, peerID)
		}
		n.peersMu.Unlock()
	}()
	for {
		select {
		case <-n.done:
			return
		default:
		}
		_, data, err := conn.ReadMessage()
		if err != nil {
			return
		}
		var env TPCPEnvelope
		if err := json.Unmarshal(data, &env); err != nil {
			log.Printf("[TPCPNode] invalid envelope from %s: %v", peerID, err)
			continue
		}

		// Replay protection: drop duplicate message IDs; clean up entries older than 300s.
		n.seenMessagesMu.Lock()
		if _, dup := n.seenMessages[env.Header.MessageID]; dup {
			n.seenMessagesMu.Unlock()
			continue
		}
		n.seenMessages[env.Header.MessageID] = time.Now().Unix()
		for id, ts := range n.seenMessages {
			if time.Now().Unix()-ts > 300 {
				delete(n.seenMessages, id)
			}
		}
		n.seenMessagesMu.Unlock()

		// Verify inbound signature if a public key is registered for the sender.
		// Fail-closed: if the message carries a signature but the sender's public key
		// is not registered, treat it as unverifiable and drop it.
		if env.Signature != "" {
			n.peerKeysMu.RLock()
			pubKey, known := n.peerKeys[env.Header.SenderID]
			n.peerKeysMu.RUnlock()
			if !known {
				log.Printf("[TPCPNode] signed message from unknown peer %s — dropping", env.Header.SenderID)
				continue
			}
			var payloadVal interface{}
			if err := json.Unmarshal(env.Payload, &payloadVal); err == nil {
				canonical, err := CanonicalJSON(payloadVal)
				if err == nil && !Verify(pubKey, canonical, env.Signature) {
					log.Printf("[TPCPNode] signature verification FAILED for message from %s — dropping", env.Header.SenderID)
					continue
				}
			}
		}

		n.dispatch(&env)
	}
}

func (n *TPCPNode) dispatch(env *TPCPEnvelope) {
	n.handlersMu.RLock()
	handler, ok := n.handlers[env.Header.Intent]
	n.handlersMu.RUnlock()
	if !ok {
		n.DLQ.Enqueue(env)
		return
	}
	handler(env)
}
